blog/2025-01-13-typescript-sdk-release.md

```diff
@@ -0,0 +1,375 @@
+
+```typescript
+export interface Main {
+  video: BlobRef
+  captions?: Caption[]
+  alt?: string
+  aspectRatio?: AppBskyEmbedDefs.AspectRatio
+  [k: string]: unknown
+}
+```
+
+Because the `$type` property is missing from that interface, developers could write invalid code, without getting an error from TypeScript:
+
+```typescript
+import { AppBskyFeedPost } from '@atproto/api'
+
+// Aliased for clarity
+type BlueskyPost = AppBskyFeedPost.Main
+
+// Invalid post, but TypeScript did not complain
+const myPost: BlueskyPost = {
+  text: 'Hey, check this out!',
+  createdAt: '2021-09-01T12:34:56Z',
+  embed: {
+    // Notice how we are missing the `$type` property here
+    video: {
+      /* reference to the video file, omitted for brevity */
+    },
+  },
+}
+```
+
+Similarly, because Bluesky post’s `embed` property was previously [typed](https://github.com/bluesky-social/atproto/blob/5ece8c6aeab9c5c3f51295d93ed6e27c3c6095c2/packages/api/src/client/types/app/bsky/feed/post.ts#L25-L31) like this:
+
+```typescript
+export interface Record {
+  // ...
+  embed?:
+    | AppBskyEmbedImages.Main
+    | AppBskyEmbedVideo.Main
+    | AppBskyEmbedExternal.Main
+    | AppBskyEmbedRecord.Main
+    | AppBskyEmbedRecordWithMedia.Main
+    | { $type: string; [k: string]: unknown }
+}
+```
+
+It was possible to create a post with a completely invalid "video" embed, and still get no error from the type system:
+
+```typescript
+import { AppBskyFeedPost } from '@atproto/api'
+
+// Aliased for clarity
+type BlueskyPost = AppBskyFeedPost.Main
+
+const myPost: BlueskyPost = {
+  text: 'Hey, check this out!',
+  createdAt: '2021-09-01T12:34:56Z',
+  embed: {
+    $type: 'app.bsky.embed.video',
+    video: 43, // This is invalid, but TypeScript does not complain
+  },
+}
+```
+
+We have fixed these issues by making the `$type` property in the generated interfaces explicit. The `app.bsky.embed.video` interface now looks like this:
+
+```typescript
+export interface Main {
+  $type?: $Type<'app.bsky.embed.video', 'main'>
+  video: BlobRef
+  captions?: Caption[]
+  alt?: string
+  aspectRatio?: AppBskyEmbedDefs.AspectRatio
+}
+```
+
+Notice how the `$type` property is defined as optional (`?:`) here. This is due to the fact that lexicons can define schemas that can be referenced from other places than open unions. In those places, there might not be any ambiguity as to how the data should be interpreted. For example, an embed that represents a "Record With Media" has a `record` property that expects an `app.bsky.embed.record` object:
+
+```typescript
+export interface Main {
+  $type?: $Type<'app.bsky.embed.recordWithMedia', 'main'>
+  record: AppBskyEmbedRecord.Main // Also used in post's `embed` property
+  media: /* omitted */
+}
+```
+
+Since there is no ambiguity as to the type of the data here, making the `$type` property required would cause a huge un-necessary bloat. Making the `$type` property optional allows to declare a "Record With Media" as follows:
```

Comment:
```suggestion
Since there is no ambiguity as to the type of the data here, making the `$type` property required would cause unnecessary bloat. Making the `$type` property optional allows to declare a "Record With Media" as follows:
```